src/debug/telemetry/telemetry.c

```diff
@@ -23,11 +28,363 @@ LOG_MODULE_DECLARE(ipc, CONFIG_SOF_LOG_LEVEL);
+	k_spinlock_key_t key;
+	int ret = sys_bitarray_alloc(bitmap->array, 1, offset);
+
+	key = k_spin_lock(&bitmap->array->lock);
+	if (!ret)
+		bitmap->occupied++;
+	k_spin_unlock(&bitmap->array->lock, key);
+	return ret;
+}
+
+static int perf_bitmap_free(struct perf_bitmap * const bitmap, size_t offset)
+{
+	k_spinlock_key_t key;
+	int ret =  sys_bitarray_free(bitmap->array, 1, offset);
+
+	key = k_spin_lock(&bitmap->array->lock);
+	if (!ret)
+		bitmap->occupied--;
+	k_spin_unlock(&bitmap->array->lock, key);
+	return ret;
+}
+
+static int perf_bitmap_setbit(struct perf_bitmap * const bitmap, size_t bit)
+{
+	return sys_bitarray_set_bit(bitmap->array, bit);
+}
+
+static int perf_bitmap_clearbit(struct perf_bitmap * const bitmap, size_t bit)
+{
+	return sys_bitarray_clear_bit(bitmap->array, bit);
+}
+
+static inline uint16_t perf_bitmap_get_occupied(struct perf_bitmap * const bitmap)
+{
+	return bitmap->occupied;
+}
+
+static inline uint16_t perf_bitmap_get_size(struct perf_bitmap * const bitmap)
+{
+	return bitmap->size;
+}
+
+static bool perf_bitmap_is_bit_clear(struct perf_bitmap * const bitmap, size_t bit)
+{
+	int val;
+	int ret = sys_bitarray_test_bit(bitmap->array, bit, &val);
+
+	if (ret < 0)
+		return false;
+	return !val;
+}
+
+struct perf_data_item_comp *perf_data_getnext(void)
+{
+	int idx;
+	int ret = perf_bitmap_alloc(&performance_data_bitmap, &idx);
+
+	if (ret < 0)
+		return NULL;
+	/* ref. FW did not set the bits, but here we do it to not have to use
+	 * isFree() check that the bitarray does not provide yet. Instead we will use isClear
+	 * ,and always set bit on bitmap alloc.
+	 */
+	ret = perf_bitmap_setbit(&performance_data_bitmap, idx);
+	if (ret < 0)
+		return NULL;
+	return &perf_data[idx];
+}
+
+int perf_data_free(struct perf_data_item_comp * const item)
+{
+	/* find index of item */
+	int idx = (item - perf_data) / sizeof(*item);
+	int ret = perf_bitmap_clearbit(&performance_data_bitmap, idx);
+
+	if (ret < 0)
+		return ret;
+	ret = perf_bitmap_free(&performance_data_bitmap, idx);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+void perf_data_item_comp_reset(struct perf_data_item_comp *perf)
+{
+	perf->total_iteration_count = 0;
+	perf->total_cycles_consumed = 0;
+	perf->restricted_total_iterations = 0;
+	perf->restricted_total_cycles = 0;
+	perf->restricted_peak_cycles = 0;
+	perf->item.peak_kcps = 0;
+	perf->item.avg_kcps = 0;
+}
 
+void perf_data_item_comp_init(struct perf_data_item_comp *perf, uint32_t resource_id,
+			      uint32_t power_mode)
+{
+	perf_data_item_comp_reset(perf);
+	perf->item.resource_id = resource_id;
+	perf->item.is_removed = false;
+	perf->item.power_mode = power_mode;
+}
+
+int get_performance_data(struct global_perf_data * const global_perf_data)
+{
+	if (!global_perf_data) {
+		tr_err(&ipc_tr, "IPC data is NULL");
+		return -EINVAL;
+	}
+
+	size_t slots_count;
+	size_t slot_idx = 0;
+	struct telemetry_wnd_data *wnd_data =
+		(struct telemetry_wnd_data *)ADSP_DW->slots[SOF_DW_TELEMETRY_SLOT];
```

Comment:
I think the file is pretty general otherwise, but this use of the memory window is Intel specific. This itself is ok, but I think it's a problem to call this "SOF debug telemetry" which suggests a generic mechanism, while in practise this only works on platforms that provide a very specific memory window structure (described in Zephyr soc/xtensa/intel_adsp/common/include/adsp-debug-window.h ).